### PR TITLE
Issue #980: Add support for Telegram Bot API Token

### LINKED ISSUE #980: Add support for Telegram Bot API Token
Add support for Telegram Bot API Token. Format is described here: https://core.telegram.org/bots/api#authorizing-your-bot

Please consoder following requirements:

- Implemented via Gitleaks Rules
- Match variable assigment
- Match environment variables assigment (.env) and YAML files
- Match URL containing token

cc @zricethezav

### PR DESCRIPTION
**Pull request by b4bay**
### Description:
This MR adds a new rule for Telegram Bot API Token.

It matches a token in the following cases:

- Variable assigment
- Constructor call
- URL containing token
- .env files
- YAML files

It closes #980

### Checklist:

- [x]  Does your PR pass tests?
- [x] Have you written new tests for your changes?
- [x] Have you lint your code locally prior to submission?

cc @zricethezav

### CODE REVIEW

File: cmd/generate/config/rules/telegram.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+			"telegram",
+			"api",
+			"bot",
+			"token",
+			"url",
+		},
+	}
+
+	// validate
+	token := secrets.NewSecret(numeric(strconv.Itoa(rand.Intn(11)+5)) + ":A" + alphaNumericExtendedShort("34"))
```
[the remark this answers is not in the document]
Comment (b4bay) [reply cmd/generate/config/rules/telegram.go:29]: math/rand is using to generate bot_id length in the test token. Seems it isn't an issue.
#jit_ignore_finding

[the remark this answers is not in the document]
Comment (weineran) [reply cmd/generate/config/rules/telegram.go:29]: Rather than using Intn(11), I'd recommend breaking this up into non-random cases to test the corner cases explicitly:

**True positives (tps)**
1. min length (5?)
2. max length (15?)

**False positives**
1. min length - 1 (4?)
2. max length + 1 (16?)

[the remark this answers is not in the document]
Comment (b4bay) [reply cmd/generate/config/rules/telegram.go]: @weineran sounds reasonable, thanks.
Done by commit #b245919

**Review by weineran — APPROVED** [review]

**Comment by weineran** [comment]
Looks good! @zricethezav can you merge if you agree?

File: cmd/generate/config/rules/telegram.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/zricethezav/gitleaks/v8/config"
+)
+
+func TelegramBotToken() *config.Rule {
+	// define rule
+	r := config.Rule{
+		Description: "Telegram Bot API Token",
+		RuleID:      "telegram-bot-api-token",
+		SecretGroup: 1,
+		Regex:       regexp.MustCompile(`(?i)(?:^|[^0-9])([0-9]{5,16}:A[a-zA-Z0-9_\-]{34})(?:$|[^a-zA-Z0-9_\-])`),
```
Comment (zricethezav) [inline cmd/generate/config/rules/telegram.go:16]: @b4bay 
I'm concerned that this doesn't follow the same patterns as the other rules. Even though you have `telegram` in the keywords, this rule could still be incorrectly associating matches since you have other strings like `api, bot, token, url`, all of which are generic. Is `:A` suppose to be the uniquely identifiable string in a telegram token?

Comment (b4bay) [reply cmd/generate/config/rules/telegram.go:16]: @zricethezav 

Your concerns are understandable.
`regexp.MustCompile` is used instead of recommended `generateSemiGenericRegex` / `generateUniqueTokenRegex` just because there is no way to match all the cases mentioned in https://github.com/zricethezav/gitleaks/pull/121#issuecomment-435955348. Or at least I wasn't lucky enough to find it.

> Is :A suppose to be the uniquely identifiable string in a telegram token?

Of course not. I bet that the coincidence of all four conditions:

1. 5-16 digits, following by
2. Exactly 35 alphanumerics
3. Starting with `A`
4. And have `telegram`, `api`, `bot`, `token` or `url` in the variable name

is unlikely.

Comment (zricethezav) [reply cmd/generate/config/rules/telegram.go:16]: @b4bay that's fair. If it turns out this is generating too many false positives we can always tune it more. If you resolve the conflicts I'll go ahead and merge

Comment (b4bay) [reply cmd/generate/config/rules/telegram.go:16]: @zricethezav resolved by 9f2379a5639483fce1a95caf9e78743e1ff3db7a
